### DISCUSSION TITLE
Multiple tasks for one worker.

### DISCUSSION THREAD

**Discussion by Alexey6486** [jasnell/piscina#151] [Q&A]
Hello.
Is there a way to add several tasks in one worker pool without using Promise.all? For instance, if we make multiple http requests and one of them is rejected then we loose all tasks in pool (finished and unfinished).

**Comment by addaleax** [comment]
@Alexey6486 This sounds a bit like a generic Promise question – but you could e.g. use `Promise.allSettled` to get the status of all Promises that you are waiting on.

**Reply by Alexey6486** [reply]
Thank you for your answer. I'll try it.
However, I still can't get how to separate tasks. Now I'm trying to compute big numbers, but all computations go sequentially.
Let's say there are 10 numbers to compute, I've got 3 unoccupied CPUs, should not it be like 3 tasks per thread?

`
    await (async function () {
      for (let i = 0; i < 10; i++) {
        const result = await piscina.run({data: i});
        console.log(result);
      }
    })();
`

What I'm doing wrong?

**Reply by addaleax** [reply]
@Alexey6486 The problem is that `await` actually waits for each task to finish first, before it allows continuing the loop. This is something where you probably actually want Promise.all:

```js
const result = await Promise.all([0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(i => piscina.run({data: i}));
```

**Reply by Alexey6486** [reply]
I tried a couple of libs and noticed that the result from worker must be awaited otherwise it returns promise in pending. It seems that the only way to work with a dynamic worker calls amount is Promise.All, otherwise, particularly for loop, it is always one thread.
